lessons/module-1/js-pure-functions.md

```diff
@@ -0,0 +1,133 @@
+---
+title: "JS: Pure Functions"
+length: 120
+tags: javascript, functions, functional programming
+---
+
+## Learning Goals
+
+* Speak to what makes a function pure and the benefits it can offer
+* Understand what side effects are and why it's important to avoid them
+* Practice transforming impure functions into more pure and predictable functions
+
+## Vocab
+
+- `Pure Function` A function that has no side effects.
+- `Side Effect` It is a state change that can be seen outside of a function call. (*such as changing a variable outside of a function's scope*)
+
+## What is a pure function?
+
+Over the last few weeks, you've begun writing multiple functions that complete different tasks and utilize parameters & arguments in different ways.  These tasks can include generating a random number, returning a new object, manipulating data, or making an update to the DOM.
+
+A **pure function** is a function that is predictable and always returns the same results or output when given the same input.  Another way of describing this is it's a function that has no side effects, (*it does not make any updates to the outside world*).  It is very likely that you have already written some *pure* functions in past projects.  Let's take a look at an example:
+
+<section class="answer">
+### Impure vs Pure Function Example
+
+**Impure Function:**
+```js
+var userEmail = "jeremiah@turing.io";
+
+function confirmEmail() {
+  console.log(`Your email address is currently ${userEmail}`);
+}
+
+confirmEmail();  // "Your email address is currently jeremiah@turing.io"
+
+userEmail = "jeremiah@turing.edu";
+confirmEmail(); // "Your email address is currently jeremiah@turing.edu"
+```
+
+**Pure Function:**
+```js
+function confirmEmail(userEmail) {
+  console.log(`Your email address is currently ${userEmail}`);
+}
+
+confirmEmail("jeremiah@turing.io") // "Your email address is currently jeremiah@turing.io"
+
+confirmEmail("jeremiah@turing.edu") // "Your email address is currently jeremiah@turing.edu"
+```
+</section>
+
+<section class="call-to-action">
+### On Your Own
+
+Based on the above examples and defintions, answer the following:
+
+1. What makes the first function impure?  What is the side effect?
+2. In the second example, what makes the function pure?  If you call the function with the same argument, will you always get the same result?
+3. What are the pros of making pure functions?  Are there any downsides?
```

Comment:
Love these questions

Reply:
Agreed! I think you're doing a lot to support them in these questions. We could just ask "which function is pure, which is impure?", but instead you're clear about which is which, so the meat of the questions become "why?" 👍